File: lib/utils/AuditsAndLogs.dart
import 'dart:typed_data';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_storage/firebase_storage.dart';
import 'package:firebase_core/firebase_core.dart';
import 'package:intl/intl.dart';
import '/models/AccountHolder.dart';
import 'package:path_provider/path_provider.dart';
import 'dart:io';
import 'dart:typed_data';

class Logger {
  static PushLog(String logMessage, String clas, String fucn) async {
    try {
      final timestamp = FieldValue.serverTimestamp();
      DateTime dateTime = DateTime.now();

      var doc = FirebaseFirestore.instance.collection("Logs").doc(clas +
          " " +
          fucn +
          " " +
          DateFormat('h:mm a').format(dateTime).toString());
      await doc.set({
        "Message": logMessage,
        "ClassName": clas,
        "FucntionName": fucn,
        'createdAt': timestamp,
      });
    } catch (e) {
      print(e);
    }
  }
}

class Auditer {
  static PushAudit(Map<String, dynamic> oldData, Map<String, dynamic> newData,
      String CollectionName) async {
    try {
      DateTime dateTime = DateTime.now();

      Map<String, dynamic> OldmodifiedMap = addPrefixToKeys(oldData, 'old');
      Map<String, dynamic> NewmodifiedMap = addPrefixToKeys(newData, 'new');
      Map<String, dynamic> AuditDoc = {...OldmodifiedMap, ...NewmodifiedMap};
      AuditDoc["CollectionName"] = CollectionName;
      var doc1 = FirebaseFirestore.instance.collection("Audits").doc(
          CollectionName +
              " " +
              DateFormat('h:mm a').format(dateTime).toString());
      await doc1.set(AuditDoc);
    } catch (e) {
      print(e);
    }
  }

  static Map<String, dynamic> addPrefixToKeys(
      Map<String, dynamic> originalMap, String prefix) {
    return Map.fromEntries(originalMap.entries
        .map((entry) => MapEntry('$prefix${entry.key}', entry.value)));
  }
}

File: Admin_Dashboard/Controllers/AuditsAndLogs.cs
using Google.Cloud.Firestore;
using Google.Cloud.Firestore.V1;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Admin_Dashboard.Controllers
{
    public class Logger
    {
        public static async void PushLog ( string logMessage, string className, string functionName )
        {
            try
            {
                FirestoreDb db = FirestoreDb.Create ("quickpall", new FirestoreClientBuilder
                {
                    CredentialsPath = "quickpall-firebase-adminsdk-wfso1-77e2bb731a.json"
                }.Build ());
                var timestamp = FieldValue.ServerTimestamp;
                DateTime dateTime = DateTime.Now;

                var doc = db
                    .Collection ("Logs")
                    .Document ($"{className} {functionName} {dateTime.ToString ("h:mm tt")}");

                await doc.SetAsync (new Dictionary<string, object>
                {
                    { "Message", logMessage },
                    { "ClassName", className },
                    { "FunctionName", functionName },
                    { "createdAt", timestamp }
                });
            }
            catch ( Exception e )
            {
                Console.WriteLine (e);
            }
        }
    }

    public class Auditer
    {
        public static async void PushAudit ( Dictionary<string, object> oldData, Dictionary<string, object> newData, string collectionName )
        {
            try
            {
                FirestoreDb db = FirestoreDb.Create ("quickpall", new FirestoreClientBuilder
                {
                    CredentialsPath = "quickpall-firebase-adminsdk-wfso1-77e2bb731a.json"
                }.Build ());
                DateTime dateTime = DateTime.Now;

                Dictionary<string, object> oldModifiedMap = AddPrefixToKeys (oldData, "old");
                Dictionary<string, object> newModifiedMap = AddPrefixToKeys (newData, "new");
                Dictionary<string, object> auditDoc = newModifiedMap.Concat (oldModifiedMap).ToDictionary (x => x.Key, x => x.Value);

                auditDoc["CollectionName"] = collectionName;

                var doc = db
                    .Collection ("Audits")
                    .Document ($"{collectionName} {dateTime.ToString ("h:mm tt")}");

                await doc.SetAsync (auditDoc);
            }
            catch ( Exception e )
            {
                Console.WriteLine (e);
            }
        }

        private static Dictionary<string, object> AddPrefixToKeys ( Dictionary<string, object> originalMap, string prefix )
        {
            return originalMap.ToDictionary (entry => $"{prefix}{entry.Key}", entry => entry.Value);
        }
    }
}
